v2/uuid/pgxcompat/uuid.go

```diff
@@ -0,0 +1,258 @@
+		}
+
+		*u = UUID{UUID: uuid.UUID(value), Status: pgtype.Present}
+	case *string:
+		if value == nil {
+			*u = UUID{Status: pgtype.Null}
+		} else {
+			return u.Set(*value)
+		}
+	default:
+		return fmt.Errorf("cannot convert %v of type %T to UUID", value, value)
+	}
+
+	return nil
+}
+
+func (u UUID) Get() interface{} {
+	switch u.Status {
+	case pgtype.Present:
+		return u.UUID
+	case pgtype.Null:
+		return nil
+	default:
+		return u.Status
+	}
+}
+
+func (u *UUID) AssignTo(target interface{}) (err error) {
+	switch u.Status {
+	case pgtype.Present:
+		switch dst := target.(type) {
+		case *uuid.UUID:
+			*dst = u.UUID
+			return nil
+		case *[16]byte:
+			*dst, err = toBinary(u.UUID.String())
+			return err
+		case *[]byte:
+			*dst = make([]byte, uuidBinaryLength)
+			bin, err := toBinary(u.UUID.String())
+			copy(*dst, bin[:])
+
+			return err
+		case *string:
+			*dst = u.UUID.String()
+			return nil
+		default:
+			if nextDst, retry := pgtype.GetAssignToDstType(dst); retry {
+				return u.AssignTo(nextDst)
+			}
+
+			return fmt.Errorf("unable to assign to %T", dst)
+		}
+	case pgtype.Null:
+		return pgtype.NullAssignTo(target)
+	}
+
+	return fmt.Errorf("cannot assign %v into %T", u, target)
+}
+
+func (u *UUID) DecodeText(ci *pgtype.ConnInfo, value []byte) error {
+	if value == nil {
+		*u = UUID{Status: pgtype.Null}
+		return nil
+	}
+
+	uuidValue := uuid.UUID(string(value))
+	if err := uuidValue.Validate(); err != nil {
+		return err
+	}
+
+	*u = UUID{UUID: uuidValue, Status: pgtype.Present}
+
+	return nil
+}
+
+func (u *UUID) DecodeBinary(ci *pgtype.ConnInfo, value []byte) error {
+	if value == nil {
+		*u = UUID{Status: pgtype.Null}
+		return nil
+	}
+
+	uuidValue, err := fromBinary(value)
+	if err != nil {
+		return err
+	}
+
+	*u = UUID{UUID: uuid.UUID(uuidValue), Status: pgtype.Present}
+
+	return nil
+}
+
+func (u UUID) EncodeText(ci *pgtype.ConnInfo, outBuf []byte) ([]byte, error) {
+	switch u.Status {
+	case pgtype.Null:
+		return nil, nil
+	case pgtype.Undefined:
+		return nil, errUndefined
+	}
+
+	return append(outBuf, u.UUID.String()...), nil
+}
+
+func (u UUID) EncodeBinary(ci *pgtype.ConnInfo, outBuf []byte) ([]byte, error) {
+	switch u.Status {
+	case pgtype.Null:
+		return nil, nil
+	case pgtype.Undefined:
+		return nil, errUndefined
+	}
+
+	tmpBuf, err := toBinary(u.UUID.String())
+	if err != nil {
+		return nil, err
+	}
+
+	return append(outBuf, tmpBuf[:]...), nil
+}
+
+// Scan implements the database/sql Scanner interface.
+func (u *UUID) Scan(value interface{}) error {
+	if value == nil {
+		*u = UUID{Status: pgtype.Null}
+		return nil
+	}
+
+	switch value := value.(type) {
+	case string:
+		return u.DecodeText(nil, []byte(value))
+	case []byte:
+		return u.DecodeText(nil, value)
+	}
+
+	return fmt.Errorf("cannot scan %T", value)
+}
+
+// Value implements the database/sql/driver Valuer interface.
+func (u UUID) Value() (driver.Value, error) {
```

Comment:
if you need pointer receiver for mutation, all should be pointer receivers.

Reply:
Why? I saw [this guy](that https://medium.com/@meeusdylan/when-to-use-pointers-in-go-44c15fe04eac) likes it, but I'm not sure I agree. This way it's more explicit that the method does not mutate the object.

Also, the blog post talks mostly about convenience for the caller not having to care, buth this method will be called by the pgx driver and not used directly by any other consumer.

Reply:
Fair enough :)